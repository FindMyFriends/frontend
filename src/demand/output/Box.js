import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Toggle from 'material-ui/Toggle';
import { requestedConfirm } from './../../ui/actions';
import { retract } from './../../demand/endpoints';
import { Link } from 'react-router-dom';
import moment from 'moment';
import { TableActionIcon, DangerIcon } from './../../components/menu/Icon';
import {
  Table,
  TableBody,
  TableHeader,
  TableHeaderColumn,
  TableRow,
  TableRowColumn,
} from 'material-ui/Table';

const Center = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

const handleRetract = (history, id) => (dispatch) => {
  dispatch(requestedConfirm(
    'Are you sure, you want to retract demand?',
    () => dispatch(retract(id, history)),
  ));
};

export const orderArrow = (sort, sorts) => {
  if (Object.prototype.hasOwnProperty.call(sorts, sort)) {
    return (
      sorts[sort][0] === '+'
        ? <i className="material-icons">arrow_drop_up</i>
        : <i className="material-icons">arrow_drop_down</i>
    );
  }
  return null;
};

const SortColumn = ({
  children, name, sorts, onSort,
}) => (
  <React.Fragment>
    <a href="#" onClick={() => onSort(name)}>
      <i className="material-icons">sort</i>
    </a>
    {children} {orderArrow(name, sorts)}
  </React.Fragment>
);

export const Box = ({ demands, dispatch, history, onSort, sorts }) => {
  if (demands.length === 0) {
    return (
      <React.Fragment>
        <Center>
          <h3>No demands</h3>
        </Center>
      </React.Fragment>
    );
  }
  return (
    <React.Fragment>
      <Table selectable={false}>
        <TableHeader displaySelectAll={false} adjustForCheckbox={false}>
          <TableRow>
            <TableHeaderColumn>
              <SortColumn
                name="id"
                sorts={sorts}
                onSort={onSort}
              >
                Position
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>
              <SortColumn
                name="general.firstname"
                sorts={sorts}
                onSort={onSort}
              >
                Firstname
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>
              <SortColumn
                name="general.lastname"
                sorts={sorts}
                onSort={onSort}
              >
                Lastname
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>
              <SortColumn
                name="general.sex"
                sorts={sorts}
                onSort={onSort}
              >
                Sex
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>
              <SortColumn
                name="general.age"
                sorts={sorts}
                onSort={onSort}
              >
                Age
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>
              <SortColumn
                name="created_at"
                sorts={sorts}
                onSort={onSort}
              >
                Created at
              </SortColumn>
            </TableHeaderColumn>
            <TableHeaderColumn>Soulmates</TableHeaderColumn>
            <TableHeaderColumn>Action</TableHeaderColumn>
          </TableRow>
        </TableHeader>
        <TableBody displayRowCheckbox={false}>
          {demands.map((demand, index) => {
            return (
              <TableRow key={demand.id}>
                <TableRowColumn>{++index}</TableRowColumn>
                <TableRowColumn>{demand.general.firstname}</TableRowColumn>
                <TableRowColumn>{demand.general.lastname}</TableRowColumn>
                <TableRowColumn>{demand.general.sex}</TableRowColumn>
                <TableRowColumn>{`${demand.general.age.from} - ${demand.general.age.to}`}</TableRowColumn>
                <TableRowColumn>{moment(demand.created_at).format('MM/DD/YYYY HH:mm')}</TableRowColumn>
                <TableRowColumn>
                  <Link to={`/demands/${demand.id}/soulmates`}>
                    {demand.soulmates.length}
                  </Link>
                </TableRowColumn>
                <TableRowColumn>
                  <span title="Visit">
                    <Link to={`/demands/${demand.id}`}>
                      <TableActionIcon className="material-icons">visibility</TableActionIcon>
                    </Link>
                  </span>
                  <span title="Delete" onClick={() => dispatch(handleRetract(history, demand.id))}>
                    <DangerIcon className="material-icons">delete</DangerIcon>
                  </span>
                </TableRowColumn>
              </TableRow>
            );
        })}
        </TableBody>
      </Table>
    </React.Fragment>
  );
};

export default Box;
