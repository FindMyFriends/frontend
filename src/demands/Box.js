import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const Box = ({ demand, more, onListing }) => {
  const moreFields = demand => {
    return lessFields(demand).concat([
      <li key="general.firstname">Firstname: {demand.general.firstname}</li>,
      <li key="general.lastname">Lastname: {demand.general.lastname}</li>,
    ]);
  };

  const lessFields = demand => {
    return [
      <li key="general.age">Age: {demand.general.age}</li>,
      <li key="general.gender">Gender: {demand.general.gender}</li>,
      <li key="general.race">Race: {demand.general.race}</li>,
    ];
  };

  return (
    <ul>
      {more ? moreFields(demand) : lessFields(demand)}
      <li>
        <Link to={`/demands/${demand.id}`}>Detail</Link>
      </li>
      <li>
        <a onClick={() => onListing(demand.id)}>{more ? 'Less' : 'More'}</a>
      </li>
    </ul>
  );
};

Box.propTypes = {
  demand: PropTypes.object.isRequired,
  more: PropTypes.bool.isRequired,
  onListing: PropTypes.func.isRequired,
};

export default Box;
