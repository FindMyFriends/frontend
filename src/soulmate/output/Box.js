import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Toggle from 'material-ui/Toggle';
import RaisedButton from 'material-ui/RaisedButton';
import {
  Table,
  TableBody,
  TableHeader,
  TableHeaderColumn,
  TableRow,
  TableRowColumn,
} from 'material-ui/Table';
import RefreshIndicator from 'material-ui/RefreshIndicator';

const yesNo = (value: mixed) => (value ? 'Yes' : 'No');

const Center = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
`;

export const RefreshButton = ({ requests, onRefresh }) => {
  const properties = {
    style: {
      position: 'relative',
      cursor: 'pointer',
      maxWidth: '50%',
    },
    size: 40,
    left: 0,
    top: 10,
  };
  const request = requests[0];
  if (request) {
    if (request.status === 'pending') {
      return (
        <Center>
          <RefreshIndicator {...properties} status="loading" />
        </Center>
      );
    } else if (request.is_refreshable) {
      return (
        <Center>
          <RefreshIndicator
              {...properties}
              onClick={onRefresh}
              percentage={100}
              status="ready"
            />
          </Center>
      );
    }
  }
  return null;
};

export const orderArrow = (sort, sorts) => {
  if (Object.prototype.hasOwnProperty.call(sorts, sort)) {
    return (
      sorts[sort][0] === '+'
        ? <i className="material-icons">arrow_drop_up</i>
        : <i className="material-icons">arrow_drop_down</i>
    );
  }
  return null;
};

export const Box = ({
  soulmates, requests, onRefresh, onClarify, onSort, sorts,
}) => {
  if (soulmates.length === 1 && soulmates[0].id === null) {
    return (
      <div>
        <p>No matches</p>
        <RefreshButton requests={requests} onRefresh={onRefresh} />
      </div>
    );
  }
  return (
    <React.Fragment>
      <RefreshButton requests={requests} onRefresh={onRefresh} />
      <Table selectable={false}>
        <TableHeader displaySelectAll={false} adjustForCheckbox={false}>
          <TableRow>
            <TableHeaderColumn>
              <a href="#" onClick={() => onSort('position')}>
                <i className="material-icons">sort</i>
              </a>
                Position {orderArrow('position', sorts)}
            </TableHeaderColumn>
            <TableHeaderColumn>
              <a href="#" onClick={() => onSort('is_correct')}>
                <i className="material-icons">sort</i>
              </a>
              Is correct {orderArrow('is_correct', sorts)}
            </TableHeaderColumn>
            <TableHeaderColumn>
              <a href="#" onClick={() => onSort('new')}>
                <i className="material-icons">sort</i>
              </a>
              Is new {orderArrow('new', sorts)}
            </TableHeaderColumn>
            <TableHeaderColumn>
              <a href="#" onClick={() => onSort('ownership')}>
                <i className="material-icons">sort</i>
              </a>
              Ownership {orderArrow('ownership', sorts)}
            </TableHeaderColumn>
            <TableHeaderColumn>
              Evolution
            </TableHeaderColumn>
          </TableRow>
        </TableHeader>
        <TableBody displayRowCheckbox={false}>
          {soulmates.map((soulmate) => {
            return (
              <TableRow key={soulmate.id}>
                <TableRowColumn>{soulmate.position}</TableRowColumn>
                <TableRowColumn>
                  <Toggle
                    toggled={soulmate.is_correct}
                    onToggle={(event, checked) => onClarify(soulmate.id, { is_correct: checked })}
                  />
                </TableRowColumn>
                <TableRowColumn>{yesNo(soulmate.new)}</TableRowColumn>
                <TableRowColumn>{soulmate.ownership}</TableRowColumn>
                <TableRowColumn>
                  <span title="Visit">
                    <a href={`${soulmate.evolution_id}`}>
                      <i className="material-icons">visibility</i>
                    </a>
                  </span>
                </TableRowColumn>
              </TableRow>
            );
        })}
        </TableBody>
      </Table>
    </React.Fragment>
  );
};

Box.propTypes = {
  soulmates: PropTypes.array.isRequired,
  requests: PropTypes.array.isRequired,
  onRefresh: PropTypes.func.isRequired,
  onClarify: PropTypes.func.isRequired,
  onSort: PropTypes.func.isRequired,
  sorts: PropTypes.object.isRequired,
};

RefreshButton.propTypes = {
  requests: PropTypes.array.isRequired,
  onRefresh: PropTypes.func.isRequired,
};

export default Box;
