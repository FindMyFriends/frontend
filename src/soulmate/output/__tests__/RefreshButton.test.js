import React from 'react';
import renderer from 'react-test-renderer';
import 'raf/polyfill';
import MuiThemeProvider from 'material-ui/styles/MuiThemeProvider';
import { RefreshButton } from '../Box';

it('ready refresh indicator for available refresh', () => {
  const tree = renderer
    .create(
      <MuiThemeProvider>
        <RefreshButton
          requests={[{ is_refreshable: true }]}
          onRefresh={() => {}}
        />
      </MuiThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('no indicator for done task', () => {
  const tree = renderer
    .create(
      <MuiThemeProvider>
        <RefreshButton
          requests={[{ is_refreshable: false, status: 'succeed' }]}
          onRefresh={() => {}}
        />
      </MuiThemeProvider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
