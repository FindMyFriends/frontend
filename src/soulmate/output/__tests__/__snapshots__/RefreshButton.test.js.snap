// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no indicator for done task 1`] = `null`;

exports[`ready refresh indicator for available refresh 1`] = `
<div
  className="sc-bdVaJa grbsvW"
>
  <div
    onClick={[Function]}
    size={40}
    style={
      Object {
        "WebkitTapHighlightColor": "rgba(0,0,0,0)",
        "backgroundColor": "#ffffff",
        "borderRadius": "50%",
        "boxShadow": "0 1px 6px rgba(0, 0, 0, 0.12),
               0 1px 4px rgba(0, 0, 0, 0.12)",
        "boxSizing": "border-box",
        "color": "rgba(0, 0, 0, 0.87)",
        "cursor": "pointer",
        "fontFamily": "Roboto, sans-serif",
        "height": 40,
        "left": -10000,
        "maxWidth": "50%",
        "muiPrepared": true,
        "opacity": 1,
        "padding": 4,
        "position": "relative",
        "top": -10000,
        "transform": "translate(10000px, 10010px)",
        "transition": "none",
        "width": 40,
        "zIndex": 2,
      }
    }
  >
    <svg
      style={
        Object {
          "height": 32,
          "width": 32,
        }
      }
      viewBox="0 0 32 32"
    >
      <circle
        cx={16}
        cy={16}
        r={11}
        style={
          Object {
            "fill": "none",
            "muiPrepared": true,
            "opacity": 1,
            "stroke": "#00bcd4",
            "strokeDasharray": "55.67600313861911, 13.439035240356333",
            "strokeDashoffset": -23.03834612632515,
            "strokeLinecap": "round",
            "strokeWidth": 3,
          }
        }
      />
      <polygon
        points="21.75,16 32.25,16 27,21.25"
        style={
          Object {
            "fill": "#00bcd4",
            "muiPrepared": true,
            "opacity": 1,
            "transform": "rotate(410deg)",
            "transformOrigin": "16px 16px",
          }
        }
      />
    </svg>
  </div>
</div>
`;
